Updated explicit backing fields proposal (#289)
Kotlin/KEEP: proposals/explicit-backing-fields.md

```kotlin
class C {
    val elementList: List<Element>
        field = mutableListOf()
}
```

## Table of contents

<!--- TOC -->

* [Use Cases](#use-cases)
  * [Expose read-only subtype](#expose-read-only-subtype)
  * [Decouple storage type from external representation](#decouple-storage-type-from-external-representation)
  * [Expose read-only view](#expose-read-only-view)
  * [Access field from outside of getter and setter](#access-field-from-outside-of-getter-and-setter)
* [Design](#design)
  * [Explicit Backing Fields](#explicit-backing-fields)
    * [Restrictions](#restrictions)
    * [Accessors](#accessors)
    * [Visibility](#visibility)
    * [Lateinit](#lateinit)
  * [Smart Type Narrowing](#smart-type-narrowing)
* [Alternatives](#alternatives)
  * [Initial Proposal](#initial-proposal)
* [Future Enhancements](#future-enhancements)
  * [Direct Backing Field Access](#direct-backing-field-access)
  * [Protected Fields](#protected-fields)
  * [Mutable Fields for Read-only Properties](#mutable-fields-for-read-only-properties)

<!--- END -->

## Use Cases

This proposal caters to a variety of use-cases that are currently met via a backing property pattern.

> ### Expose read-only subtype

Author: Thanks. Fixed.

* Rx `Observable` has a mutable `Subject` counterpart.
* Kotlin coroutines `SharedFlow` has a `MutableSharedFlow`, etc.

For example, sample code [from an Android app](https://github.com/elpassion/crweather/blob/9c3e3cb803b7e4fffbb010ff085ac56645c9774d/app/src/main/java/com/elpassion/crweather/MainModel.kt#L14):

```kotlin
private val mutableCity = MutableLiveData<String>().apply { value = "" }
private val mutableCharts = MutableLiveData<List<Chart>>().apply { value = emptyList() }
private val mutableLoading = MutableLiveData<Boolean>().apply { value = false }
private val mutableMessage = MutableLiveData<String>()

val city: LiveData<String> = mutableCity
val charts: LiveData<List<Chart>> = mutableCharts
val loading: LiveData<Boolean> = mutableLoading
val message: LiveData<String> = mutableMessage
```

Becomes:

```kotlin
val city: LiveData<String>
    field = MutableLiveData().apply { value = "" }
val charts: LiveData<List<Chart>>
    field = MutableLiveData().apply { value = emptyList() }
val loading: LiveData<Boolean>
    field = MutableLiveData().apply { value = false }
val message: LiveData<String>
    field = MutableLiveData()

> ```

Reviewer: It can be noted that while the total number of lines stays roughly the same, the related properties and fields are better grouped together, and also the repetition in property names is avoided.

Author: Added a note to that effect.

but is seen as a property with read-only type to the outside code.

### Decouple storage type from external representation

Sometimes a property must be internally represented by a different type for storage-efficiency or architectural reason,
while having a different outside type. For example, API requirements might dictate that the property type
is `String`, but if we know that it always represents a decimal integer, then it can be efficiently stored as such 
with custom getter and custom setter.

```kotlin
val number: String
    field: Int = 0
    get() = field.toString()
    set(value) { field = value.toInt() }
```

### Expose read-only view

In some application it is desirable to expose not just a read-only subtype, but a specially constructed read-only
view that protects the data structure from casting into mutable type. For example a `MutableStateFlow` has
[`asStateFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-state-flow.html)
extension for that purpose.

With the proposed syntax it is possible to declare a custom getter:

```kotlin
val state: StateFlow<State> 

>     get() = field.asStateFlow()

Reviewer: This creates a view on each getter invocation which may be unwanted.

Author: Added clarification on this point.

    field = MutableStateFlow(State.INITIAL)
```

For this use-case, the TBD syntax of [Direct Backing Field Access](#direct-backing-field-access) will need to be added.

### Access field from outside of getter and setter

Kotlin allows property field to be accessed from the property's getter and setter using a `field` variable.
This proposal is designed with an idea to provide an explicit syntax to access a property's field from anywhere
inside the corresponding class when the field is declared explicitly:

```kotlin
class Component {
    var status: Status
        field // explicit field declaration
        set(value) {
            field = value
            notifyStatusChanged()
        }
}
```

> This way, all the code inside the class can change the field of the property directly, without invoking the setter . 

Reviewer: ```suggestion
This way, all the code inside the class can change the field of the property directly, without invoking the setter. 
```

_Statistics_: found in 3.7k files

Primary property also might be `protected` or `internal`
([example](https://github.com/meganz/android/blob/21a8dd8da80dfdf1dfed58ad1734a2c76fadac6f/app/src/main/java/mega/privacy/android/app/presentation/logout/LogoutViewModel.kt#L27-L28)).

### Delegation

It's possible that either backing property ([example](https://github.com/b-lam/Resplash/blob/4b13d31134d1c31bd331e92ffe8d410984529212/app/src/main/java/com/b_lam/resplash/ui/upgrade/UpgradeViewModel.kt#L32-L40))
or primary property ([example](https://github.com/google/accompanist/blob/0d0198d0cab599295f7601cb386d5989d72cc8cd/pager/src/main/java/com/google/accompanist/pager/PagerState.kt#L145))
is delegated.

_Statistics_: backing property delegation found in 2693 files, primary property delegation in 1010 files.

## Design

### Explicit Backing Fields

By analogy with the already existing term ["implicit backing fields"](https://kotlinlang.org/docs/properties.html#backing-properties), it is proposed to introduce a syntax
for declaring "explicit backing fields" of properties.

> Syntax is consisted of a keyword `field`, optional type definition and initialization expression

Reviewer 2: ```suggestion
Syntax consists of a keyword `field`, optional type definition and initialization expression
```

```kotlin
private val _city = MutableLiveData<String>()
val city: LiveData<String> get() = _city
```

Could be rewritten as follows:

```kotlin
val city: LiveData<String>
    field = MutableLiveData()
```

### Visibility

`private` is the default and the only allowed visibility for explicit backing fields
(supporting other visibilities might be considered, see [Future enhancements](#future-enhancements)).

Visibility of property must be more permissive than explicit backing field visibility.

### Resolution

Calls of properties with explicit backing field are resolved to
- the backing field, if call is made from inside the class (within private scope),
- getter, otherwise.

```kotlin
class SomeViewModel {
    val city: LiveData<String>
        field = MutableLiveData("")
    
    fun updateCity(newCity: String) {
        city.value = newCity // visible as MutableLiveData, calling field
    }
}

fun outside(vm: SomeViewModel) {
    vm.city // visible as LiveData, calling getter
} 
```

There is no possibility to call a getter inside the class
(might be reconsidered in [Future enhancements](#future-enhancements)).

### Accessors

> Property, which has explicit backing field, can also specify getter or setter.

Reviewer 2: ```suggestion
Property, which has explicit backing field, can also specify getter and/or setter.
```

The following additional restrictions apply:

* It's prohibited to reference `field` inside explicit backing field initializer expression.
* Explicit backing field must be referenced through `field` in property initializer, setter and getter (to prevent feature abuse).

### Type inference

Explicit type declaration on explicit backing field is optional. If it is not specified, it's inferred from the explicit backing field initialization expression.

Explicit type declaration on property, which has explicit backing field, is optional. If it is not specified, it's inferred from
the property initialization or from getter. If none of them are specified, the property type is equal to the explicit backing field type.

It's a warning if the property type is equal to the explicit backing field type.

### Other restrictions

There are the following additional semantic restrictions on the property declaration.
Property with explicit backing field:

* must be final (otherwise calling a field instead of a getter would be undesirable behavior)
* can't be `const`
* can't be `expect` or `external`

> * can't be `inline` (but its accessors can)

Reviewer 2: FTR: an `inline` property means its accessors are `inline`.

Why do we want to have `inline` accessors for EBF properties? This means that the EBF leaks into the public ABI, has to be accessible from the outside and I'm not sure I like this.

Atm, `inline` properties cannot have backing fields, and I do not understand why we want to relax the restriction here.

Reviewer 2: Smth I originally forgot -- do we say anywhere that EBF are prohibited for extension properties? =)

> * or explicit backing field itself can't be `lateinit` (added to [Future enhancements](#future-enhancements)).

Reviewer 2: Why "or"?

}

fun outside(vm: SomeViewModel) {
    vm.city // visible as LiveData, calling getter
} 
```

There is no possibility to call a getter inside the class
(might be reconsidered in [Future enhancements](#future-enhancements)).

### Accessors

Property, which has explicit backing field, can also specify getter and/or setter.

```kotlin
val counter: Int
    field = atomic(0)
    get() = field.value
```

When explicit backing field with type `F` for a property with type `P` is declared,
then compiler can derive getter and, for `var` properties, setter implementation automatically:

* If `F` is a subtype of `P`, the compiler can provide a default getter.
* If `P` is a subtype of `F`, the compiler can provide a default setter.

If the compiler can not provide a getter, the user must declare it explicitly.
The same applies to setters in case of `var` properties.

```kotlin
val city: String
    field = MutableLiveData<String>()
    get() = field.value ?: "-" // It is an error if getter is not explicitly specified here
```

### Explicit backing field in combination with property initializer

It's allowed to have both property initializer and explicit backing field specified.

```kotlin

> val city: StateFlow<String> = field.asStateFlow()

Reviewer 3: What about cases this case:
```kotlin
val city: StateFlow<String> get() = field.asStateFlow()
    field = MutableStateFlow("")
```

Will it be supported? Now it's not really clear for me from the KEEP text

Reviewer 3: Ah, indeed. Thanks

```kotlin
private val _itemCount = atomic(0)
val itemCount: Int get() = _itemCount.value
```

It is also possible with other containers:

- Other atomic classes ([example](https://github.com/Sorapointa/Sorapointa/blob/31578cb9e460bb4516b51fcef4f0c4af5600caa9/sorapointa-core/src/main/kotlin/org/sorapointa/game/PlayerAvatarComp.kt#L42-L82))
- `StateFlow`, `LiveData`, etc ([example](https://github.com/c5inco/Compose-Modifiers-Playground/blob/2c7e8c55eeaab15e38696397a84b57475a84a6a3/ideaPlugin/src/main/kotlin/intellij/SwingColors.kt#L44-L52))
- `ThreadLocal` ([example](https://github.com/corda/corda/blob/a95b854b1eda8a1e9284c0db8ff43b78ea0eb290/core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt#L56))
- `WeakReference` ([example](https://github.com/expo/expo/blob/a502f90d6d9345648157aceb16a78f044669842e/packages/expo-dev-menu/android/src/main/java/expo/modules/devmenu/devtools/DevMenuDevToolsDelegate.kt#L21-L32))

#### Access backing property inside

It is worth emphasizing that in most cases the primary property is only for public use and is not
referenced within the class. Obviously, there is no point in doing so, since the backing property usually provides

> all the same functionality, as a primary property and even more.

Reviewer 4: Please don't use the word "obviously" in such a context. This is not a conclusion within a set of strict constraints, but the rationale behind a free choice we decide to make

or primary property ([example](https://github.com/google/accompanist/blob/0d0198d0cab599295f7601cb386d5989d72cc8cd/pager/src/main/java/com/google/accompanist/pager/PagerState.kt#L145))
is delegated.

_Statistics_: backing property delegation found in 2693 files, primary property delegation in 1010 files.

## Design

### Explicit Backing Fields

By analogy with the already existing term "implicit backing fields", it is proposed to introduce a syntax
for declaring "explicit backing fields" of properties.
Syntax is consisted of a keyword `field`, optional type definition and initialization expression
(more detailed [Grammar changes](#grammar-changes) are in the separate section).

```kotlin
private val _city = MutableLiveData<String>()
val city: LiveData<String> get() = _city
```

Could be rewritten as follows:

```kotlin
val city: LiveData<String>
    field = MutableLiveData()
```

### Visibility

`private` is the default and the only allowed visibility for explicit backing fields
(supporting other visibilities might be considered, see [Future enhancements](#future-enhancements)).

> Visibility of property can be any, except `private`.

Reviewer 4: Minor: let's say "must be more permissive than", because this statement would stay correct w.r.t. possible future enhancements

```kotlin
val counter: Int
    field = atomic(0)
    get() = field.value
```

When explicit backing field with type `F` for a property with type `P` is declared,
then compiler can derive getter and, for `var` properties, setter implementation automatically:

* If `P :> F`, the compiler can provide a default getter.
* If `P <: F`, the compiler can provide a default setter.

If the compiler can not provide a getter, the user must declare it explicitly.
The same applies to setters in case of `var` properties.

```kotlin
val city: String
    field = MutableLiveData<String>()
    get() = field.value ?: "-" // It is an error if getter is not explicitly specified here
```

### Explicit backing field in combination with property initializer

It's allowed to have both property initializer and explicit backing field specified.

```kotlin
val city: StateFlow<String> = field.asStateFlow()
    field = MutableStateFlow("")
```

In this case, the property initializer expression (`field.asStateFlow()` in example above) is evaluated once
immediately after the backing field is initialized and then returned every time the getter of property is called.

> If property with explicit backing field has initializer, it's prohibited to declare accessors.

Reviewer 4: One more candidate for the future enhancements list? Or a way "to prevent feature abuse"?

     | (field? (semi? getter)? (semi? setter)?) 
     | (field? (semi? setter)? (semi? getter)?))

field ::=
    modifiers? 'field' (':' type)? '=' expression     
```

### Java interop

For JVM target, property with explicit backing field is compiled into field(-s) and accessors.
Property can be accessed from Java through them in accordance with the following visibility rules:

* accessors are assigned visibility of property
* field is assigned `private` visibility of explicit backing field 

```kotlin
protected val name: LiveData<String>
    field = MutableLiveData<String>()
```

So, the property above is visible in Java as:

```java
private final MutableLiveData<String> name = MutableLiveData();

protected final LiveData<String> getName() {
    return this.name;
}
```

`@JvmField` is prohibited on properties with explicit backing field, because in this case a field type becomes exposed,
which makes use of the feature completely pointless.

#### Property with initializer compilation

Property with explicit backing field and initializer is represented by additional
auxiliary synthetic field which is needed for storing the value from property initializer.

> Auxiliary field is given the name `propertyName$init` and is `private` regardless of property visibility.

Reviewer 4: It may be worth mentioning if allowing everything (accessors + access to both the fields as `field` and `init` within them) is considered a future enhancement or something we'd like to avoid for design reasons

On JVM backing field can be obtained using [`javaField` extension](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect.jvm/java-field.html).

There is no API in Reflection to check whether property has an explicit backing field or obtain any information about it.
Adding such an API might be considered in [Future enhancements](#future-enhancements).

## Future enhancements

1. Make it possible to access getter instead of explicit backing field when referenced within the scope of visibility of explicit backing field. This is not supported because usually there is no need for calling getter (see [Access backing property inside](#access-backing-property-inside)).
2. Support other visibilities of explicit backing field (`protected` or `internal`).
3. Support delegation of explicit backing field or property.
4. Support `lateinit` explicit backing field.
5. Support combining mutable explicit backing field and non-mutable property. 
Despite its popularity (see [`var` backing property](#var-backing-property)), this use case is not supported in this proposal, because it is impossible for one property to behave as mutable + nullable and
at the same time non-mutable and non-nullable.

> 6. Add API in Reflection to retrieve information about explicit backing field.

Reviewer 4: Since this KEEP also suggests bringing in the `init` field, it can probably raise a similar concern?

    get() = field.toString()
    set(value) { field = value.toInt() }
```

### Expose read-only view

In some application it is desirable to expose not just a read-only subtype, but a specially constructed read-only
view that protects the data structure from casting into mutable type. For example a `MutableStateFlow` has
[`asStateFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-state-flow.html)
extension for that purpose.

With the proposed syntax it is possible to declare a custom getter:

```kotlin
val state: StateFlow<State> 
    get() = field.asStateFlow()
    field = MutableStateFlow(State.INITIAL)
```

For this use-case, the TBD syntax of [Direct Backing Field Access](#direct-backing-field-access) will need to be added.

### Access field from outside of getter and setter

Kotlin allows property field to be accessed from the property's getter and setter using a `field` variable.
This proposal is designed with an idea to provide an explicit syntax to access a property's field from anywhere
inside the corresponding class when the field is declared explicitly:

```kotlin
class Component {
    var status: Status

>         field // explicit field declaration

Reviewer 5: This line is redundant as it is

Reviewer 5: Also, there is no initializer here

Author: Let's discuss this part of design here. Do we automatically make backing fields of _all properties_ available inside the class via TBD syntax or only if you explicitly declare a backing field for them?

Reviewer 5: I think this is better

> we automatically make backing fields of all properties available inside the class via TBD syntax

Author: Ok. Corrected.
